examples/asr/asr_chunked_inference/rnnt/speech_to_text_streaming_infer_rnnt.py

```diff
@@ -0,0 +1,470 @@
+    encoder_frame2audio_samples = features_frame2audio_samples * model_stride
+
+    context_encoder_frames = ContextSize(
+        left=int(cfg.left_context_secs * features_per_sec / model_stride),
+        chunk=int(cfg.chunk_secs * features_per_sec / model_stride),
+        right=int(cfg.right_context_secs * features_per_sec / model_stride),
+    )
+    context_samples = ContextSize(
+        left=context_encoder_frames.left * model_stride * features_frame2audio_samples,
+        chunk=context_encoder_frames.chunk * model_stride * features_frame2audio_samples,
+        right=context_encoder_frames.right * model_stride * features_frame2audio_samples,
+    )
+
+    full_ctx_audio_samples = context_samples.total()
+    logging.info(
+        "Corrected contexts (sec): "
+        f"Left {context_samples.left / audio_sample_rate:.2f}, "
+        f"Chunk {context_samples.chunk / audio_sample_rate:.2f}, "
+        f"Right {context_samples.right / audio_sample_rate:.2f}"
+    )
+    logging.info(f"Corrected contexts (subsampled encoder frames): {context_encoder_frames}")
+    logging.info(f"Corrected contexts (in audio samples): {context_samples}")
+    latency_secs = (context_samples.chunk + context_samples.right) / audio_sample_rate
+    logging.info(f"Theoretical latency: {latency_secs:.2f} seconds")
+
+    with torch.no_grad(), torch.inference_mode():
+        all_hyps = []
+        for i in tqdm(range(0, len(records), cfg.batch_size)):
+            # get audio
+            audio_batch, audio_batch_lengths = get_audio_batch(
+                [record["audio_filepath"] for record in records[i : i + cfg.batch_size]],
+                device=map_location,
+                sample_rate=audio_sample_rate,
+            )
+            batch_size = audio_batch.shape[0]
+            device = audio_batch.device
+
+            # decode audio by chunks
+
+            current_hyps = None
+            state: Optional[BatchedGreedyDecodingState] = None
+            left_sample = 0
+            # right_sample = initial latency in audio samples
+            right_sample = min(context_samples.chunk + context_samples.right, audio_batch.shape[1])
+            # start with empty buffer
+            buffer = torch.zeros([batch_size, 0], dtype=audio_batch.dtype, device=device)
+            buffer_size = ContextSize(left=0, chunk=0, right=0)
+            buffer_size_batch = ContextSizeBatch(
+                left=torch.zeros_like(audio_batch_lengths),
+                chunk=torch.zeros_like(audio_batch_lengths),
+                right=torch.zeros_like(audio_batch_lengths),
+            )
+            rest_audio_lengths = audio_batch_lengths.clone()
+
+            # iterate over audio samples
+            while left_sample < audio_batch.shape[1]:
+                # add samples to buffer
+                buffer = torch.cat((buffer, audio_batch[:, left_sample:right_sample]), dim=1)
```

Comment:
maybe also create same fixed empty buffer and update each time to not create new tensor everytime?

Reply:
Let's postpone this optimization, since the overhead from buffer manipulation is tiny compared to CPU-GPU moves for hypotheses. It will make the code more complicated, especially if we want to avoid left padding (which hurts the recognition quality).